### DISCUSSION TITLE
Ruff linter + autofixer

### DISCUSSION THREAD

**Discussion by MattF-NSIDC** [nsidc/tech-radar#21] [Tools]
Tried [ruff](https://github.com/charliermarsh/ruff) for the first time today and it really impressed me. It's got a lot of value over `flake8` + plugins aside from just speed (10-100x faster). It supports a [huge number of linting rules](https://beta.ruff.rs/docs/rules/) from across the _whole ecosystem_, not just `flake8`. It autofixes basically everything, which eliminates the need for `isort` as a separate step. In combination with [pre-commit](https://nsidc.slack.com/archives/C4UEHRWKC/p1684643455728559), linting and formatting takes no effort, and is portable without a dependency on end-user editor setup. I give `ruff` 5 bones out of 5 :bone::bone: :bone::bone: :bone:

Configuration is great. Integrates with `pyproject.toml`. Can turn on/off categories of rules as well as individual rules, and can set which ones will autofix.

**Comment by MattF-NSIDC** [comment]
Due to the number of rules, autofix capability, and `pyproject.toml` support, (and the speed helps, too!) I think we should consider adopting `ruff` as a replacement for `flake8`. Setting up ruff + black with pre-commit and letting everything get autofixed before hitting the repo is an absolute dream :heart:

**Reply by mikedorfman** [reply]
Good to know - I've never heard of Ruff and this seems like it might be a helpful replacement for flake8 in our Cumulus lambda linting.

**Comment by flamingbear** [comment]
I looked at this and it's cool. But I think this should probably be under linting / code-fixing #16 (do those two things go together? I think so)

**Reply by MattF-NSIDC** [reply]
I'm struggling with when to create a card for a specific tech and when to create a card for categories. Seems like we have places for both. "Tools" for things like "ruff" or "flake8", and "techniques" for things like "linting / code-fixing"? Once we agree as a team on a clear prescription for this, I'd like to document that decision in the repo!

**Reply by flamingbear** [reply]
Generally I want to avoid specific libraries where possible as cards.  I thought we discussed that we would use comments under the topics and then mark one as the correct answer.  This presumes we can change that answer later, no idea if that's true or not.

**Reply by MattF-NSIDC** [reply]
Answer can be changed after-the-fact

**Reply by MattF-NSIDC** [reply]
>I thought we discussed that we would use comments under the topics and then mark one as the correct answer.

I think we're not following this pattern so far? For example, we have Terraform and IaC cards both created in the last hour.

**Reply by flamingbear** [reply]
That's probably because we just moved stuff over from the trello board during the programmers meeting
